classes/controller/class-block-search.php

```diff
@@ -0,0 +1,264 @@
+ * Block search
+ *
+ * @package P4BKS\Controllers
+ */
+
+namespace P4GBKS\Controllers;
+
+use wpdb;
+
+/**
+ * Block search in post content
+ *
+ * Usage:
+ *   $search = new Search ( $wpdb );
+ *   // All blocks with a className attribute in block options
+ *   $search->find( 'className' );
+ *   // All blocks of type planet4-blocks/covers
+ *   $search->find( null, ['block_type' => 'planet4-blocks/covers'] );
+ *   // All blocks sorted by post_title
+ *   $search->find( null, null, [ 'post_title' ] );
+ */
+class Block_Search {
+
+	/**
+	 * @var wpdb
+	 */
+	private $db;
+
+	/**
+	 * @var array[]
+	 */
+	private $results = [];
+
+	/**
+	 * @var int
+	 */
+	private $block_count = 0;
+
+	/**
+	 * @var int
+	 */
+	private $post_count = 0;
+
+	/**
+	 * @param wpdb $wpdb WP DB object.
+	 */
+	public function __construct( wpdb $wpdb ) {
+		$this->db = $wpdb;
+	}
+
+	/**
+	 * @param ?string   $search  Search in blocks names and options.
+	 * @param ?array    $filters Filter by exact dimension value.
+	 * @param ?string[] $sort    Sort by dimension (natural sort).
+	 *
+	 * @return array[] Array of parsed items
+	 * Item format:
+	 * [
+	 *   'post_id'       => int,
+	 *   'post_title'    => string,
+	 *   'post_status'   => string,
+	 *   'post_type'     => string,
+	 *   'post_date'     => string (datetime),
+	 *   'post_modified' => string (datetime),
+	 *   'guid'          => string,
+	 *   'block_ns'      => string,
+	 *   'block_type'    => string,
+	 *   'block_opts'    => string (json),
+	 *   'match'         => string,
+	 * ]
+	 */
+	public function find(
+		?string $search = null,
+		?array $filters = null,
+		?array $sort = null
+	): array {
+		$this->reset();
+
+		$query = $this->get_query( $search );
+		$res   = $this->db->get_results( $query );
+		$items = [];
+		foreach ( $res as $post ) {
+			$items = array_merge( $items, $this->parse_post( $post ) );
+		}
+
+		$filtered = $this->filter_items( $items, $filters, $search );
+		$sorted   = $this->sort_items( $filtered, $sort );
+
+		$this->results     = $sorted;
+		$this->block_count = count( $this->results );
+		$this->post_count  = count( array_unique( array_column( $this->results, 'post_id' ) ) );
+
+		return $this->results;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function results(): array {
+		return $this->results;
+	}
+
+	/**
+	 * Number of posts in items.
+	 *
+	 * @return int
+	 */
+	public function post_count(): int {
+		return $this->post_count;
+	}
+
+	/**
+	 * Number of blocks in items.
+	 *
+	 * @return int
+	 */
+	public function block_count(): int {
+		return $this->block_count;
+	}
+
+	/**
+	 * @param ?string $search Search string.
+	 * @return string
+	 * @throws \UnexpectedValueException Empty prepared query.
+	 */
+	private function get_query( ?string $search ) {
+		$regex = empty( $search )
+			? '.*<!-- wp:.* /?-->.*'
+			: '.*<!-- wp:.*' . $search . '.* /?-->.*';
+
+		$query = $this->db->prepare(
+			"SELECT
+				ID, post_content, post_title, post_status,
```

Comment:
If we use it for purging, we only need to have the ID field, else the result set could get quite big since it's fetching all matches with no limit.

Reply:
Ideally we could get the URL to purge directly in the query results, but I think that's a bit tricky given there are many things that need to be checked to know the actual URL.